### PR TITLE
[SPARK-47357][SQL] Add support for Upper, Lower, InitCap (all collations)

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Addition of InitCap support and tests for Upper, Lower, InitCap.

### Why are the changes needed?
We need to support functions with collations.

### Does this PR introduce _any_ user-facing change?
Yes, we enabled more functions.

### How was this patch tested?
Test in `CollationStringExpressionsSuite`.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 1]
```diff
@@ -89,6 +89,45 @@ class CollationStringExpressionsSuite
     testRepeat("UNICODE_CI", 3, "abc", 2)
   }
 
+  test("UPPER check output type on collated string") {
```
Comment (nikolamand-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:92]: Maybe it's more readable to combine multiple simple checks for expressions like this into one test function like [here](https://github.com/apache/spark/pull/45933/files#diff-ab481fd0c9704381d8535339727e08673da432080e47be49ab002411adef2afa). In this case we could also combine `testUpper`, `testLower`, `testInitCap` into one function.

**Review by nikolamand-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 1 above)
Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:168]: Yeah, agreed. Will refactor a bit.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 2]
```diff
@@ -163,6 +165,45 @@ class CollationStringExpressionsSuite
...
+      checkEvaluation(Collation(Lower(s)).replacement, expected)
+    }
+
+    testLower("UTF8_BINARY", 0, "abc")
+    testLower("UTF8_BINARY_LCASE", 1, "abc")
+    testLower("UNICODE", 2, "abc")
+    testLower("UNICODE_CI", 3, "abc")
+  }
+
+  test("INITCAP check output type on collated string") {
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:194]: @uros-db @nikolamand-db @dbatomic @stefankandic @stevomitric @cloud-fan @MaxGekk  I am a bit sceptical about the recent change we introduced in this Suite with UTF8String refactoring. Why are we opting to e2e tests for each expression? Correct me if I am wrong, but these expressions are proven to work e2e with existing tests. Shouldn't we write only unit tests for each expression to test whether these "black boxes" work correctly under different types of collations and reserve e2e tests for multiple stuff, like correctly working from parsing through analysis to execution.

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:194]: The line of reasoning is to cover both unit tests and e2e tests:
1. first use [CollationSupportSuite.java](https://github.com/apache/spark/blob/master/common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java) with a lot of rigorous tests to see whether collation-aware implementation introduced in `CollationSupport.scala` is correct
2. then use [CollationStringExpressionsSuite.scala](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala) with just a couple of SQL tests to verify whether full queries with collation-aware expressions in `stringExpressions.scala` work as expected

I tried to boil this down in: https://issues.apache.org/jira/browse/SPARK-47410

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:192]: It's a good practice to always have end-to-end tests for each function. The test coverage doesn't need to be very high (the unit tests should cover all the corner cases), but we should have one.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**